File: play_error_sound.py
import socket
import signal
import sys
import os
import time
from datetime import datetime
import subprocess
import logging
from logging.handlers import RotatingFileHandler

class ErrorSoundServer:
    """
    the client:
    import socket
    import logging

    def play_error_sound(message):
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.settimeout(1)  # 1秒超时
                sock.connect(('10.211.55.2', 9123))
                sock.send(message.encode('utf-8'))
            return True
        except Exception as e:
            logging.error(f"Server playErrorSound connection failed: {e}")
            return False
            
    usage: python ~/Coding/myscript/play_error_sound.py 1 >/dev/null 2>&1 &
    """
    def __init__(self):
        self.config = {
            'port': 9123,
            'log_file': os.path.join(os.path.dirname(__file__), 'play_error_sound.log'),
            'max_log_size': 10 * 1024 * 1024,  # 10MB
            'max_log_backup': 3,
            'enable_sound': True,
            'sound_file': '/System/Library/Sounds/Ping.aiff',
            'buffer_size': 2048,
            'pid_file': os.path.join(os.path.dirname(__file__), 'play_error_sound.pid'),
        }
        
        self.setup_logging()
        self.sock = None
        self.setup_signal_handlers()
        self.write_pid_file()

    def setup_logging(self):
        handler = RotatingFileHandler(
            self.config['log_file'],
            maxBytes=self.config['max_log_size'],
            backupCount=self.config['max_log_backup']
        )
        formatter = logging.Formatter('%(asctime)s - %(message)s')
        handler.setFormatter(formatter)
        
        self.logger = logging.getLogger('ErrorSoundServer')
        self.logger.setLevel(logging.INFO)
        self.logger.addHandler(handler)

    def setup_signal_handlers(self):
        signal.signal(signal.SIGTERM, self.signal_handler)
        signal.signal(signal.SIGINT, self.signal_handler)
        signal.signal(signal.SIGHUP, self.signal_handler)

    def signal_handler(self, signum, frame):
        self.cleanup()
        sys.exit(0)

    def write_pid_file(self):
        with open(self.config['pid_file'], 'w') as f:
            f.write(str(os.getpid()))

    def cleanup(self):
        if self.sock:
            self.sock.close()
        try:
            os.remove(self.config['pid_file'])
        except OSError:
            pass

    def play_sound(self):
        if self.config['enable_sound']:
            try:
                subprocess.Popen(['afplay', self.config['sound_file']], 
                               stdout=subprocess.DEVNULL, 
                               stderr=subprocess.DEVNULL)
            except Exception as e:
                self.logger.error(f"Failed to play sound: {e}")

    def run(self):
        try:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.sock.bind(('0.0.0.0', self.config['port']))
            self.sock.listen(5)
            self.sock.setblocking(False)
            
            self.logger.info(f"Listening on port {self.config['port']}...")
            
            while True:
                try:
                    conn, addr = self.sock.accept()
                    try:
                        data = conn.recv(self.config['buffer_size'])
                        if data:
                            message = data.decode('utf-8').strip()
                            self.logger.info(message)
                            self.play_sound()
                    finally:
                        conn.close()
                except BlockingIOError:
                    time.sleep(0.1)  # 避免CPU占用过高
                except Exception as e:
                    self.logger.error(f"Error handling connection: {e}")

        except Exception as e:
            self.logger.error(f"Server error: {e}")
            self.cleanup()
            sys.exit(1)

def main():
    server = ErrorSoundServer()
    server.run()

if __name__ == '__main__':
    main()

File: play_error_sound.php
<?php
/*
the client
function playErrorSound($message) {
    $errno = 0;
    $errstr = '';
    $fp = @fsockopen('10.211.55.2', 9123, $errno, $errstr, 1);
    if (!$fp) {
        LcLog::error("Server playErrorSound connection failed: [$errno] $errstr");
        return false;
    }
    fwrite($fp, $message);
    fclose($fp);
    return true;
}

usage: php ~/Coding/myscript/play_error_sound.php
kill: kill `cat ~/Coding/myscript/play_error_sound.pid`
status: ps aux | grep play_error_sound
or cat ~/Coding/myscript/play_error_sound.pid
*/

declare(ticks = 1);
date_default_timezone_set('Asia/Shanghai');

// 配置项
$config = [
    'port' => 9123,
    'logFile' => __DIR__ . '/play_error_sound.log',
    'maxLogSize' => 10 * 1024 * 1024,
    'maxLines' => 10000,
    'cleanStrategy' => 'size',
    'enableSound' => true,
    'soundFile' => '/System/Library/Sounds/Ping.aiff',
    'bufferSize' => 2048,
    'timeFormat' => 'Y-m-d H:i:s',
    'pidFile' => __DIR__ . '/play_error_sound.pid',
];

// 守护进程化
function daemonize() {
    // 1. 创建子进程
    $pid = pcntl_fork();
    if ($pid == -1) {
        die("Error: Can't fork process\n");
    } else if ($pid > 0) {
        // 父进程退出
        exit(0);
    }

    // 2. 设置新会话组长
    if (posix_setsid() == -1) {
        die("Error: Can't setsid()\n");
    }

    // 3. 再次fork以确保进程不是会话组长
    $pid = pcntl_fork();
    if ($pid == -1) {
        die("Error: Can't fork process\n");
    } else if ($pid > 0) {
        exit(0);
    }

    // 4. 修改工作目录
    chdir('/');

    // 5. 重设文件掩码
    umask(0);

    // 6. 关闭标准输入输出和错误输出
    fclose(STDIN);
    fclose(STDOUT);
    fclose(STDERR);

    // 重新打开标准输入输出和错误输出到/dev/null
    $stdIn = fopen('/dev/null', 'r');
    $stdOut = fopen('/dev/null', 'w');
    $stdErr = fopen('/dev/null', 'w');
}

// 启动守护进程
daemonize();

// 信号处理
pcntl_signal(SIGTERM, 'signalHandler');
pcntl_signal(SIGHUP,  'signalHandler');
pcntl_signal(SIGINT,  'signalHandler');

function signalHandler($signal) {
    global $sock, $config;
    switch ($signal) {
        case SIGTERM:
        case SIGINT:
        case SIGHUP:
            @socket_close($sock);
            @unlink($config['pidFile']);
            exit(0);
    }
}

// 写入PID文件
file_put_contents($config['pidFile'], getmypid());

// 创建socket
$sock = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
if (!$sock) {
    error_log("Failed to create socket\n", 3, $config['logFile']);
    exit(1);
}

// 设置socket选项
socket_set_option($sock, SOL_SOCKET, SO_REUSEADDR, 1);
if (!socket_bind($sock, '0.0.0.0', $config['port'])) {
    error_log("Failed to bind socket\n", 3, $config['logFile']);
    exit(1);
}

if (!socket_listen($sock)) {
    error_log("Failed to listen socket\n", 3, $config['logFile']);
    exit(1);
}

error_log("Listening on port {$config['port']}...\n", 3, $config['logFile']);

// 设置非阻塞模式
socket_set_nonblock($sock);

while (true) {
    $conn = @socket_accept($sock);
    if ($conn) {
        $buf = '';
        socket_recv($conn, $buf, $config['bufferSize'], 0);
        
        if (!empty($buf)) {
            $logMessage = '[' . date($config['timeFormat']) . '] ' . trim($buf) . "\n";
            
            // 根据策略清理日志
            if ($config['cleanStrategy'] === 'size') {
                if (file_exists($config['logFile']) && filesize($config['logFile']) > $config['maxLogSize']) {
                    file_put_contents($config['logFile'], '');
                }
                file_put_contents($config['logFile'], $logMessage, FILE_APPEND);
            } else {
                if (file_exists($config['logFile'])) {
                    $lines = file($config['logFile']);
                    if (count($lines) >= $config['maxLines']) {
                        $lines = array_slice($lines, -(($config['maxLines'] - 1)));
                        $lines[] = $logMessage;
                        file_put_contents($config['logFile'], implode('', $lines));
                    } else {
                        file_put_contents($config['logFile'], $logMessage, FILE_APPEND);
                    }
                } else {
                    file_put_contents($config['logFile'], $logMessage, FILE_APPEND);
                }
            }
            
            // 播放声音
            if ($config['enableSound']) {
                shell_exec('afplay ' . $config['soundFile'] . ' > /dev/null 2>&1 &');
            }
        }
        
        socket_close($conn);
    }
    
    // 避免CPU占用过高
    usleep(100000); // 休眠0.1秒
}
